lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+
+use enso_prelude::*;
+
+use enso_prelude::fmt::Formatter;
+use enso_web::performance;
+use inflector::Inflector;
+use ordered_float::OrderedFloat;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_plain::from_str;
+use web_sys::PerformanceEntry;
+
+
+
+// =================
+// === Log Level ===
+// =================
+
+#[derive(Copy, Clone, Debug)]
+#[derive(Serialize, Deserialize)]
+#[allow(missing_docs)]
+pub enum ProfilingLevel {
+    Section,
+    Task,
+    Detail,
+    Debug,
+}
+
+
+// ===============
+// === Loggers ===
+// ===============
+
+/// Define a new boolean variable whose value is determined by a feature flag in the crate.
+/// The name of the variable is `ENABLED` and it will be true if a feature flag
+/// `enable-<log_level_name>-profiling` is set, otherwise it will be false.
+macro_rules! define_profiling_toggle {
+    ($log_level_name:ident) => {
+        paste::item! {
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(feature = "enable-" $log_level_name "-profiling")]
+            pub const ENABLED: bool = true;
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(not(feature = "enable-" $log_level_name  "-profiling"))]
+            pub const ENABLED: bool = false;
+        }
+    };
+}
+
+/// Define a new profiling module that exposes `start`, `end` and `measure` methods. The profiling
+/// can be activated and de-activated via a crate feature flag named
+/// `enable-<profiling_module_name>-profiling`, which will turn the profiling methods into no-ops.
+macro_rules! define_logger {
+    ($log_level:expr, $log_level_name_upper:ident, $log_level_name:ident) => {
+        /// Profiler module that exposes methods to measure named intervals.
+        pub mod $log_level_name {
+            use super::*;
+
+            define_profiling_toggle!($log_level_name);
+
+            /// Start measuring a named time interval. Return an `IntervalHandle` that can be sued
+            /// to end the profiling.
+            pub fn start(interval_name: &str) -> IntervalHandle {
+                start_interval($log_level, interval_name)
+            }
+
+            /// Manually end measuring a named time interval.
+            pub fn end(interval_name: &str) {
+                warn_on_error(end_interval($log_level, interval_name));
+            }
+
+            /// Profile the execution of the given closure.
+            pub fn measure<T, F: FnMut() -> T>(
+                interval_name: &str,
+                closure: F,
+            ) -> IntervalMeasurementResult<T> {
+                measure_interval($log_level, interval_name, closure)
+            }
+        }
+    };
+}
+
+define_logger!(ProfilingLevel::Section, SECTION, section);
+define_logger!(ProfilingLevel::Task, TASK, task);
+define_logger!(ProfilingLevel::Detail, DETAIL, detail);
+define_logger!(ProfilingLevel::Debug, DEBUG, debug);
```

Comment:
We are repeating the names here again. Instead of that, you should have a macro like `with_all_profiling_levels!(define_logger);`, which should expand to the above. The macro `with_all_profiling_levels` can be generated by procedural macro, automatically from the JS config file that is also read by typescript.

Reply:
(see above)